crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -292,95 +315,327 @@ where
-                         was modified",
-                    ),
-                ));
-            }
-        };
-
-        let pre_masp_balance: Amount = self
-            .ctx
-            .read_pre(&balance_key(&token, &Address::Internal(Masp)))?
-            .unwrap_or_default();
-        let post_masp_balance: Amount = self
-            .ctx
-            .read_post(&balance_key(&token, &Address::Internal(Masp)))?
-            .unwrap_or_default();
-        let (amount, source, target) =
-            match pre_masp_balance.cmp(&post_masp_balance) {
-                Ordering::Equal => {
-                    return Err(Error::NativeVpError(
-                        native_vp::Error::SimpleMessage(
-                            "Found a MASP transaction that moves no \
-                             transparent funds",
-                        ),
-                    ));
-                }
-                Ordering::Less => (
-                    post_masp_balance - pre_masp_balance,
-                    counterpart,
-                    Address::Internal(Masp),
-                ),
-                Ordering::Greater => (
-                    pre_masp_balance - post_masp_balance,
-                    Address::Internal(Masp),
-                    counterpart,
+        for (token, _) in masp_balances {
+            // NOTE: no need to extract the changes of the masp balances too,
+            // we'll examine those of the other transparent addresses and the
+            // multitoken vp ensures a correct match between the two sets
+            result.masp.insert(token);
+        }
+
+        for (token, counterpart) in counterparts_balances {
+            let counterpart_balance_key = match counterpart {
+                ShieldedActionOwner::Owner(addr) => balance_key(token, addr),
+                ShieldedActionOwner::Minted => minted_balance_key(token),
+            };
+            let pre_balance: Amount = self
+                .ctx
+                .read_pre(&counterpart_balance_key)?
+                .unwrap_or_default();
+            let post_balance: Amount = self
+                .ctx
+                .read_post(&counterpart_balance_key)?
+                .unwrap_or_default();
+            // Public keys must be the hash of the sources/targets
+            let address_hash = <[u8; 20]>::from(ripemd::Ripemd160::digest(
+                sha2::Sha256::digest(
+                    &counterpart.to_address_ref().serialize_to_vec(),
                 ),
+            ));
+            let mut diff = match post_balance.checked_sub(pre_balance) {
+                Some(diff) => DeltaBalance::Positive(diff),
+                None => DeltaBalance::Negative(pre_balance - post_balance),
             };
 
-        Ok(TransparentTransferData {
-            source,
-            target,
-            token,
-            amount,
-        })
+            if let ShieldedActionOwner::Minted = counterpart {
```

Comment:
I think this could be removed if we only validated the absolute value of the balance changes instead of the value and the sing. I'm not sure though if this is completely safe to do

Reply:
To my understating the new receive a shielded action two things happen:

1. We mint the token amount in the `#Multitoken/{token_addr}/balance/minted` balance (this should be the case only if we are receiving a foreign token, if we are receiving back the native token than we should unescrow from `#Multitoken/{token_addr}/balance/#Ibc`)
2. The amount is shielded, thus increasing the transparent balance of the MASP, but this does not reduce the amount of the minted balance, so in this case the sum of the amounts does not collapse to 0, leading to the need to mock an opposite balance change for the minted balance to make the computation work